### PR TITLE
feat: add component metadata for hooks

### LINKED ISSUE #4570: Support component hooks
### `snapcraft.yaml` syntax

Support the following syntax:
```yaml
components:
  <component-name>:
    hooks:
      <hook-name>:
        plugs:
          - <plug-name>
        <all-currently-valid-hooks>
```

Supported hooks are:
- install
- pre-refresh
- post-refresh
- remove

### `meta/snap.yaml` syntax

The `meta/snap.yaml` needs to list all component metadata, including the component hooks.  Currently, this is not happening.

### Why it needs to get done

To support component hooks

### PR DESCRIPTION
**Pull request by mr-cal**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `tox run -m lint`?
- [x] Have you successfully run `tox run -e test-py310`? (supported versions: `py39`, `py310`, `py311`, `py312`)

-----

Support metadata for component hooks.

This includes the project definition and the output in the component's `meta/component.yaml` file.

Built on #4701, so only the last 3 commits are relevant.

Fixes #4570
(CRAFT-2463)

### CODE REVIEW

**Review by sergiusens — COMMENTED** [review]
looks good, as we discussed it is missing additions to our components spread test, but also a unit tests in tests/unit/services/test_package.py similar to the one you have here in tests/unit/meta/test_snap_yaml.py

File: tests/spread/core22/components/task.yaml [hunk 1]
```diff
@@ -11,13 +11,19 @@ restore: |
...
-  # assert contents of default partition
+  # assert snap contents
   unsquashfs -dest "snap-contents" "hello-components_1.0_amd64.snap"
   if [ ! -e "snap-contents/bin/hello" ]; then
     echo "Expected 'bin/hello' in snap contents"
     exit 1
   fi
 
+  # assert snap metadata
+  if ! diff snap-contents/meta/snap.yaml expected-snap.yaml; then
```
Comment (sergiusens) [inline tests/spread/core22/components/task.yaml:22]: we are doing diff -u in other places (grep the code), might be good to stay consistent here

**Review by sergiusens — APPROVED** [review]
I am not sure what the point is of repeating the metadata or how plugs and slots will come into play here, but this looks technically good

**Review by tigarmo — APPROVED** [review]

File: tests/spread/core22/components/task.yaml (hunk 1 above)
Comment (mr-cal) [reply tests/spread/core22/components/task.yaml]: Added.

**Comment by mr-cal** [comment]
Will merge after #4701 lands

**Comment by mr-cal** [comment]
Force-pushing to target `main` instead of `work/CRAFT-2628-upload-components`

File: tests/spread/core22/components/expected-snap.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+confinement: strict
+grade: devel
+environment:
+  LD_LIBRARY_PATH: ${SNAP_LIBRARY_PATH}${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}
+  PATH: $SNAP/usr/sbin:$SNAP/usr/bin:$SNAP/sbin:$SNAP/bin:$PATH
+components:
+  share:
+    summary: Hello World
+    description: Hello World
+    type: test
```
Comment (sergiusens) [inline tests/spread/core22/components/expected-snap.yaml:20]: it looks like no hooks were added here

**Comment by mr-cal** [comment]
`integration-spread-tests-store` will fail with `Error in components validation: Additional properties are not allowed ('hooks' was unexpected)`
until https://code.launchpad.net/~jslarraz/review-tools/+git/review-tools/+merge/462423 is released

**Comment by mr-cal** [comment]
image

This only took me 100 million commits or so

**Comment by mr-cal** [comment]
I'm getting odd `review-tool` rejections for core24 snaps with components, so I will not upload a core24 snap in the spread tests (for now).

File: tests/spread/core22/components/expected-snap.yaml (hunk 2 above)
Comment (mr-cal) [reply tests/spread/core22/components/expected-snap.yaml:20]: Added

**Pull request merged**